Feature/autoconfig parse radio caps

tomereli opened the pull request:
Add a new API to the database to set the supported operating class as received from the Radio Basic Capabilities TLV as part of the autoconfig WSC message from the agent.
Currently, prplMesh stores channels and not operating classes. Since EasyMesh specification is using the operating classes and not directly channels, it makes sense to change our database
and channel selection code to use operating classes.
However, this will be a big change and should be done as part of the full channel selection flow.
In order to put things in the right direction, this commit adds a new API for storing the supported operating class in the DB, which performs the conversion to sWifiChannel which is used across the code. This will be used in a followup commit to storing the operating classes in
the DB which will be then updating the radio capabilities, which eventually allow bml_conn_map to show the correct band based on the operating class.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1516,10 +1513,160 @@ bool master_thread::handle_intel_slave_join(Socket *sd, ieee1905_1::CmduMessageR
-                                                std::shared_ptr<ieee1905_1::tlvWscM1> tlvwscM1,
-                                                std::string bridge_mac, std::string radio_mac,
-                                                ieee1905_1::CmduMessageTx &cmdu_tx)
+static inline std::vector<uint8_t> operating_class_to_channel_set(uint8_t operating_class)
```
Let's assume we would have wanted this function. It certainly wouldn't be here. it will be on son_wireless_utils.
Second, operating class has also bw property, which you ignored. Nitpickers would add band and dfs.

tomereli replied:
You are right, I will move it to son_wireless_utils. What do you mean by "let's assume we would want this function"? Its not matter of wanting it, I needed it for getting the right channel and understanding weather it is a 5g or 2.4g radio.

Let me remind you that this is still draft, I didn't explain why I did everything I did in this commit so try to keep an open mind:)

I ignored everything other than channel on purpose - all we need at the moment is the channel.

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1516,10 +1513,160 @@ bool master_thread::handle_intel_slave_join(Socket *sd, ieee1905_1::CmduMessageR
+        for (auto c : channel_set) {
+            beerocks_message::sWifiChannel ch;
+            ch.channel = c;
+            supported_channels.push_back(ch);
```
This part of this flow should be implemented as part of the channel selection full flow.
Besides, the fact that we have used channels list until now, does not say we have to continue working that way.
AFAIU, the Multi-AP messages use operating class instead of a channel, so maybe it will be more convenient to save it as it, instead of converting it back and forth.
IMO, print only what you get from the CMDU, and don't store it in the database. When we will implement channel selection we will know better.

tomereli replied:
The only reason I did that is for the Conn map to show if the radio is 5ghz or 2.4ghz. I agree that we should store this as is without converting but we will do this refactoring as part of channel selection.
In the meanwhile it won't hurt to have this conversion even if it is only for the network map, since it's a nice thing to show Ina demo.
In fact, it will also allow us to use our steering algorithm before implementing full channel selection which is also nice to have and we plan to try to get this working for the demo.

Commit:
database: add_hostap_supported_operating_class API

Add a new API to the database to set the supported operating class as
received from the Radio Basic Capabilities TLV as part of the autoconfig
WSC message from the agent. Currently, prplMesh stores channels and not
operating classes. Since EasyMesh specification is using the operating
classes and not directly channels, it makes sense to change our database
and channel selection code to use operating classes.

However, this will be a big change, and should be done as part of the
full channel selection flow.

In order to put things in the right direction, this commit adds a new
API for storing the supported operating class in the DB, which performs
the conversion to sWifiChannel which is used across the code.

This will be used in a followup commit to store the operating classes in
the DB which will be then update the radio capabilities, which
eventually allow bml_conn_map to show the correct band based on the
operating class.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
bcl: wireless: operating class to channel set API

Add an API for converting a global operating class to a channel set,
based table 4-E in the ieee 802.11 specification.
This will be used to convert the operating classes received in the WSC
autoconfig radio basic capabilities TLV to a channel set, which will
allow storing supported channels in the controller database.
Note that this however will be changed in a future refactor as part of
the full channel selection flow implementation which will store the
operating class directly in the database.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1520,14 +1513,198 @@ bool master_thread::handle_intel_slave_join(Socket *sd, ieee1905_1::CmduMessageR
+            return supported_channels;
+        }
+        auto &op_class_channels_rx = std::get<1>(operating_class_tuple);
+        auto operating_class       = op_class_channels_rx.operating_class();
```
I agree with Moran that it's better to store the operating classes in the database. However, I also agree with Tomer that this is too big a change. So as middle ground, I propose to already add the function to the database to set operating classes, and do the conversion of operating class to channel inside the database.

For this new function, I think it's better to use `add_hostap_supported_operating_class`, not `set_hostap_supported_operating_classes` like is done for the channels (`set_hostap_supported_channels`). That will probably make it easier to separate the parsing from the database function and the database function from how it's internally stored.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1516,10 +1513,160 @@ bool master_thread::handle_intel_slave_join(Socket *sd, ieee1905_1::CmduMessageR
-                                                ieee1905_1::CmduMessageTx &cmdu_tx)
+static inline std::vector<uint8_t> operating_class_to_channel_set(uint8_t operating_class)
+{
+    switch (operating_class) {
```
Please add a comment (either here or in the doxygen header, but probably here) that explains where you get this table.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1516,10 +1513,160 @@ bool master_thread::handle_intel_slave_join(Socket *sd, ieee1905_1::CmduMessageR
+{
+    switch (operating_class) {
+    case 81:
+        return std::vector<uint8_t>{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13};
```
Instead of this big switch, isn't it better to use a map?
```
std::map<uint8_t, std::vector<uint8_t>> operating_class_to_channel_list = {
    {81, {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13}},
...
};
```

tomereli replied:
Done

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1232,6 +1232,48 @@ const std::vector<beerocks_message::sWifiChannel> db::get_hostap_supported_chann
 
+/**
+ * @brief Add supported operating class to the database.
+ * Currently this function is a wrapper which converts the operating
```
There should be an empty line after the @brief, otherwise the whole paragraph is considered the "brief". (nitpick)

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1232,6 +1232,48 @@ const std::vector<beerocks_message::sWifiChannel> db::get_hostap_supported_chann
+ * class to a set of supported channels and updates the list of currently
+ * supported channels.
+ * 
+ * @param mac radio mac
```
In EasyMesh terminology, it would be Radio UID. (nitpcik)

morantr replied:
radio_mac and radio_uid are two different things and not necessarily equal. The current database convention is to use mac. However, I don't know how could we support it, since I don't know if we are getting the radio mac from somewhere.
Maybe you need to use here the conversion function that convert the al_mac and ruid to database key.

tomereli replied:
See `handle_non_intel_slave_join` - this is the RUID. We keep the same convention of calling it radio mac only for alignment with the current code base, but @arnout 's comment is correct so I'll add it to the description.

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1232,6 +1232,48 @@ const std::vector<beerocks_message::sWifiChannel> db::get_hostap_supported_chann
+ * @return true on success
+ * @return false on failure
+ */
+bool db::add_hostap_supported_operating_class(std::string mac, wfa_map::cOperatingClassesInfo &info)
```
So, this is *not* what I meant :-/

I would prefer to not expose the details about how things happen to be structured in the MAP TLVs to the database. So I would expand the structure into an operating class and transmit power argument and a vector of statically non-operable channels.

(nitpick)

morantr replied:
I think both the input argument should be a const reference.

morantr replied:
I think it is better to explicitly call the argument "radio_mac" instead of "mac".

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1232,6 +1232,48 @@ const std::vector<beerocks_message::sWifiChannel> db::get_hostap_supported_chann
+        beerocks_message::sWifiChannel ch;
+        ch.channel        = c;
+        ch.tx_pow         = info.maximum_transmit_power_dbm();
+        ch.radar_affected = affected_channels.find(c) != affected_channels.end();
```
??? What do *statically* non-operable channels have to do with radar?

I think statically non-operable channels should just be removed from channel_set.

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1232,6 +1232,48 @@ const std::vector<beerocks_message::sWifiChannel> db::get_hostap_supported_chann
+            std::find_if(supported_channels.begin(), supported_channels.end(),
+                         [&](const beerocks_message::sWifiChannel ch) { return ch.channel == c; });
+        if (old != supported_channels.end())
+            supported_channels.erase(old);
```
Well, no... we should keep the old one and update it with the new information - basically, setting tx_pow to the maximum of the old one and the new one.

tomereli replied:
Why the maximum of the old and new? I changed this to set the tx_pow to the new one.

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1232,6 +1232,48 @@ const std::vector<beerocks_message::sWifiChannel> db::get_hostap_supported_chann
+        // First, delete old channel if exists
+        auto old =
+            std::find_if(supported_channels.begin(), supported_channels.end(),
+                         [&](const beerocks_message::sWifiChannel ch) { return ch.channel == c; });
```
Might be worth to make channel_set a map instead of a set.

tomereli replied:
I don't understand why - Do you mean it will be simpler if instead of set we use a 1 to 1 mapping - something like a lookup table? so instead of `find_if` we can simply access `channel_map[channel]` to see if it exists?

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1232,6 +1232,48 @@ const std::vector<beerocks_message::sWifiChannel> db::get_hostap_supported_chann
+        supported_channels.push_back(ch);
+    }
+
+    set_hostap_supported_channels(mac, &supported_channels[0], supported_channels.size());
```
Instead of getting the channels, adding one, and then setting it, I think it's worthwhile to add an update function to the database, `add_hostap_supported_channels`. This would also do the merging in case it's needed.

tomereli replied:
Perhaps in a future refactor. I already spent too much time on this side quest so I rather keep this one as is (assuming nitpick - correct me if I'm wrong)

arnout commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -454,4 +454,59 @@ std::vector<uint8_t> wireless_utils::calc_5g_20MHz_subband_channels(
+        {81, {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13}},
+        {82, {14}},
+        {83, {1, 2, 3, 4, 5, 6, 7, 8, 9}},
+        {85, {}},
```
Do you know what this even means? Why do they define operating classes without channels?

tomereli replied:
These operating classes are defined as "GeoDB" - they should be used based on the regulatory domain. I will set these as invalid operating class and add a TODO comment. This will be part of the channel selection full flow.

arnout commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -454,4 +454,59 @@ std::vector<uint8_t> wireless_utils::calc_5g_20MHz_subband_channels(
+ */
+std::unordered_set<uint8_t> wireless_utils::operating_class_to_channel_set(uint8_t operating_class)
+{
+    static const std::map<uint8_t, std::unordered_set<uint8_t>> operating_class_to_channel_set = {
```
I should have thought of this before, of course, but: an operating class is more than just a channel set, it also puts limits, including bandwidth. So maybe it would be more appropriate to use sWifiChannel directly?

But that can be done in a later commit.

tomereli replied:
Part of the channel selection full flow then.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1653,6 +1699,7 @@ bool master_thread::handle_non_intel_slave_join(Socket *sd,
     //            }
     //        } else {
     database.set_hostap_band_capability(radio_mac, beerocks::SUBBAND_CAPABILITY_UNKNOWN);
+    autoconfig_wsc_parse_radio_caps(radio_mac, radio_caps);
```
This call should be a little higher, replacing the "TODO database.set_hostap_supported_channels"

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1527,10 +1524,59 @@ bool master_thread::handle_intel_slave_join(Socket *sd, ieee1905_1::CmduMessageR
+ * @return false on failure
+ */
+bool master_thread::autoconfig_wsc_parse_radio_caps(
+    std::string radio_mac, std::shared_ptr<wfa_map::tlvApRadioBasicCapabilities> radio_caps)
```
In one of my PRs, I passed this stuff by const ref instead of passing the shared_ptr, i.e. dereferencing it in the caller. There was no discussion about that, but maybe this is something we should also agree on: prefer to pass by ref or by pointer (in case the pointer is known to be non-null, obviously).

morantr replied:
I think i is better to pass is a reference, that way you don't have to do pointer validation (which you didn't do).
Also, radio_mac should be const reference.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1527,10 +1524,59 @@ bool master_thread::handle_intel_slave_join(Socket *sd, ieee1905_1::CmduMessageR
+            // add comma if not last channel in the list, else close list by add curl brackets
+            ss << (((ch_idx + 1) != non_oper_channels_list_length) ? "," : "}");
+        }
+        LOG(DEBUG) << ss.str();
```
Is it really worth adding all this code for a single debug message?

tomereli replied:
Now it is, that I changed the API to receive statically non operable channels.

arnout reviewed (approved):
As usual, a lot of comments, but all of them are either nitpicks or something that can be changed later (when we have a better idea of what is really needed).

tomereli commented:
Do not merge yet. Review however is welcome

Commit:
db: add get_hostap_supported_channels_string()

Add a function to get the supported channels in a string for debug
logging, which will be used when updating supported channels in the
controller.
This function also exists in the BWL, so the correct thing to do is to
move this function from BWL to BCL and use the same one here and there -
but this will require the BCL to depend on BTLVF which will cause
circular dependency, so instead simply implement it for the DB
seperately.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: autoconfig: add real operating class

Update JOIN_MASTER state to use a real operating class for the radio
basic capabilities (operating class 81 from Table 4-E in the ieee 80211
specification).

This will be used by a followup commit which will add the supported
operating class to the supported channels list in the controller
database.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
controller: autoconfig: parse radio capabilities

Add autoconfig_wsc_parse_radio_caps() function to master thread which
goes through the radio basic capabilities TLV and updates the database's
supported channels for the radio.

This is an initial implementation which will be finalized in the channel
selection full implementation, hence not all parameters of sWiFiChannel
are set (the idea is to change the controller and agents to use
operating class in the database instead of a list of supported channels).

This commit also updates Intel and non-Intel agents handling to parse
the received radio capabilities and update the DB.

This is meant to achieve one goal - lay the infrastructure for channel
selection implementation (nice to have) but more importantly provide
the support for non-Intel agents bands to be indentified correctly by
the prplMesh controller.

The outcome of this commit is that the non-Intel radio agents, provided
M1 contains the radio capabilities, will be identified as either 2.4GHz
or 5GHz radios.

Testing
- GW+Repeater using test_gw_repeater.sh
- testbed pass autoconfig and see that bml_conn_map shows the correct
radio bands for each radio (TODO)

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1527,10 +1524,59 @@ bool master_thread::handle_intel_slave_join(Socket *sd, ieee1905_1::CmduMessageR
-                                                std::string bridge_mac, std::string radio_mac,
-                                                ieee1905_1::CmduMessageTx &cmdu_tx)
+/**
+ * @brief parse the radio basic capabilities TLV and store the operating class
```
You forgot to use a capital letter in the sentence first letter.

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1527,10 +1524,59 @@ bool master_thread::handle_intel_slave_join(Socket *sd, ieee1905_1::CmduMessageR
+        auto operating_class            = op_class.operating_class();
+        auto maximum_transmit_power_dbm = op_class.maximum_transmit_power_dbm();
+        ss << "operating_class=" << int(operating_class);
+        ss << "maximum_transmit_power_dbm=" << int(maximum_transmit_power_dbm);
```
You forgot to add ", " before "maximum_transmit_power_dbm".

tomereli replied:
Fixed (I added newline instead)

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1527,10 +1524,59 @@ bool master_thread::handle_intel_slave_join(Socket *sd, ieee1905_1::CmduMessageR
+    }
+    for (int oc_idx = 0; oc_idx < operating_classes_list_length; oc_idx++) {
+        std::stringstream ss;
+        auto operating_class_tuple = radio_caps->operating_classes_info_list(oc_idx);
```
Maybe you can to get the pointer to the list once from the tuple, and then tread it as array (access with square brackets).
BTW, we should really get rid of this tuple, and return pointer instead.

tomereli replied:
I'm not sure I understood your proposal. In any case, I guess this should have been a (nitpick) - you're welcome to change this on a followup PR.

arnout replied:
> I'm not sure I understood your proposal. In any case, I guess this should have been a (nitpick) - you're welcome to change this on a followup PR.

Get rid of this tuple -> tlvf should generate code that returns a pointer instead of a tuple, so you don't need all this crazy std::get<1> stuff.

The part about "get the pointer to the list once from the tuple" is something I don't understand either.

morantr commented on controller/src/beerocks/master/son_master_thread.h:
```diff
@@ -54,9 +55,10 @@ class master_thread : public beerocks::socket_thread {
+    bool handle_non_intel_slave_join(
+        Socket *sd, std::shared_ptr<wfa_map::tlvApRadioBasicCapabilities> radio_caps,
+        std::shared_ptr<ieee1905_1::tlvWscM1> tlvwscM1, std::string bridge_mac,
+        std::string radio_mac, ieee1905_1::CmduMessageTx &cmdu_tx);
```
Here as well need to change strings arguments to const reference, and maybe as Arnout has suggested, to pass reference instead of shared_ptr.

tomereli replied:
This should be in a separate commit / PR and should be a (nitpick)

morantr commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -454,4 +454,59 @@ std::vector<uint8_t> wireless_utils::calc_5g_20MHz_subband_channels(
+ */
+std::unordered_set<uint8_t> wireless_utils::operating_class_to_channel_set(uint8_t operating_class)
+{
+    static const std::map<uint8_t, std::unordered_set<uint8_t>> operating_class_to_channel_set = {
```
I wonder why did you prefer unordered_set instead of set. Since there is no intention to add new values, I think there is only gain by having the channels ordered.

tomereli replied:
good point. There was no real reason to prefer unordered set on top of a normal set.

morantr commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -454,4 +454,59 @@ std::vector<uint8_t> wireless_utils::calc_5g_20MHz_subband_channels(
+ * @param operating_class operating class
+ * @return std::unordered_set<uint8_t> set of supported channels by the operating class or empty if failure
+ */
+std::unordered_set<uint8_t> wireless_utils::operating_class_to_channel_set(uint8_t operating_class)
```
const reference return

tomereli replied:
On error we return an empty set, and I don't want to define a static const empty set just for returning const reference. So I'll skip this suggestion (and should have been (nitpick))

morantr commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -454,4 +454,59 @@ std::vector<uint8_t> wireless_utils::calc_5g_20MHz_subband_channels(
+        {127, {153, 161}},
+        {180, {1, 2, 3, 4, 5, 6}}};
+
+    auto channel_set = operating_class_to_channel_set.find(operating_class);
```
It is not channel_set what you get. It is an iterator to map.
I would change the name to "op_class_it" or "just "it".
```cpp
    auto it = operating_class_to_channel_set.find(operating_class);
    if (it == operating_class_to_channel_set.end()) {
       // code
    }
    auto&  channel_set = it->second;
    return channel_set;
```
nitpick

tomereli replied:
changed name to `it`

morantr commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -454,4 +454,59 @@ std::vector<uint8_t> wireless_utils::calc_5g_20MHz_subband_channels(
+
+    auto channel_set = operating_class_to_channel_set.find(operating_class);
+    if (channel_set == operating_class_to_channel_set.end()) {
+        LOG(ERROR) << "reserved operating class " << operating_class;
```
since operating_class is uint8_t, then it needs to be casted int

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1232,6 +1232,48 @@ const std::vector<beerocks_message::sWifiChannel> db::get_hostap_supported_chann
+bool db::add_hostap_supported_operating_class(std::string mac, wfa_map::cOperatingClassesInfo &info)
+{
+    auto supported_channels = get_hostap_supported_channels(mac);
+    auto channel_set = wireless_utils::operating_class_to_channel_set(info.operating_class());
```
here also `auto&`

tomereli replied:
same answer as before, we don't return const reference on purpose.

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1232,6 +1232,48 @@ const std::vector<beerocks_message::sWifiChannel> db::get_hostap_supported_chann
+        // First, delete old channel if exists
+        auto old =
+            std::find_if(supported_channels.begin(), supported_channels.end(),
+                         [&](const beerocks_message::sWifiChannel ch) { return ch.channel == c; });
```
It is better to remove the from the lambda the reference for all, and explicitly supply what you need, and as well use the input argument as referencfe too:
```cpp
std::find_if(supported_channels.begin(), supported_channels.end(),
                         [&c](const beerocks_message::sWifiChannel& ch) { return ch.channel == c; });
```

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1232,6 +1232,48 @@ const std::vector<beerocks_message::sWifiChannel> db::get_hostap_supported_chann
+ * supported channels.
+ * 
+ * @param mac radio mac
+ * @param info operating class info from the tlvApRadioBasicCapabilities TLV
```
Capital letter on the first letter of description.

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1232,6 +1232,48 @@ const std::vector<beerocks_message::sWifiChannel> db::get_hostap_supported_chann
+ */
+bool db::add_hostap_supported_operating_class(std::string mac, wfa_map::cOperatingClassesInfo &info)
+{
+    auto supported_channels = get_hostap_supported_channels(mac);
```
what if "mac" does not exist and the function return empty container?

tomereli replied:
What is the problem in that? supported channels will be empty and it will be filled with the channels in we get according to the operating class.

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1552,24 +1555,29 @@ bool master_thread::autoconfig_wsc_parse_radio_caps(
+        ss << "operating_class=" << int(operating_class) << std::endl;
+        ss << "maximum_transmit_power_dbm=" << int(maximum_transmit_power_dbm) << std::endl;
+        ss << "channel list={ ";
+        auto channel_list = son::wireless_utils::operating_class_to_channel_set(operating_class);
```
auto&

tomereli replied:
auto& where?

morantr replied:
auto& channel_list

tomereli replied:
`operating_class_to_channel_set` doesn't return reference (and it shouldn't)

Commit:
bwl: dummy: add dummy channel by default

Add channel 1 or 149 for 2.4G or 5G radios respectively for bml_conn_map
to show correct bandwidth for radios.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1247,6 +1247,60 @@ std::string db::get_hostap_supported_channels_string(std::string mac)
+    for (auto c : non_operable_channels) {
+        auto channel =
+            std::find_if(supported_channels.begin(), supported_channels.end(),
+                         [&c](const beerocks_message::sWifiChannel ch) { return ch.channel == c; });
```
Need reference beerocks_message::sWifiChannel as well

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1247,6 +1247,60 @@ std::string db::get_hostap_supported_channels_string(std::string mac)
+    for (auto c : channel_set) {
+        auto channel =
+            std::find_if(supported_channels.begin(), supported_channels.end(),
+                         [&c](const beerocks_message::sWifiChannel ch) { return ch.channel == c; });
```
Need reference beerocks_message::sWifiChannel as well

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1232,6 +1232,21 @@ const std::vector<beerocks_message::sWifiChannel> db::get_hostap_supported_chann
     return n->hostap->supported_channels;
 }
 
+std::string db::get_hostap_supported_channels_string(std::string mac)
```
const std::string &radio_mac

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1232,6 +1232,21 @@ const std::vector<beerocks_message::sWifiChannel> db::get_hostap_supported_chann
+{
+    std::ostringstream os;
+    auto supported_channels = get_hostap_supported_channels(mac);
+    for (auto val : supported_channels) {
```
const auto val&

arnout commented:
> radio_mac and radio_uid are two different things and not necessarily equal.

Is it useful to make that distinction? AFAIU the multi-AP specification requires the radio UID to be globally unique, which implies it really is the radio MAC address. I think the standard uses UID rather than MAC address because an AP doesn't really have a single MAC address; it has a number of addresses that can be used as BSSID.

That said, for the time being it's probably better to make the distinction, until the controller code has been cleaned up to use radio UID everywhere.
